README.md
# Snowflake
## 🚨 WARNINGS 🚨
This package is only compatible with 64-bit installations of PHP.

This package has not been tested against incredibly strenuous loads such as the actual traffic seen
regularly by Twitter.

## Introduction
This is a PHP implementation of Twitter's Snowflake, Sony's Sonyflake, and our own custom
Modelflake ID generator.

Traits are also included which allow you to directly and automatically generate IDs on Eloquent Model classes.

## Installation
To add this library into your project, run:
```
composer require moves/snowflake
```

## Usage
### Eloquent Models
#### Prerequisites
For each separate deployment of your application, be sure to set a unique value for the `SNOWFLAKE_MACHINE_ID`
environment key.

The default implementation of the provided Eloquent Snowflake generator traits requires you to have a configured
Cache driver. 

🚨 **It is important to note that the `file` Cache driver is not supported for deployments on distributed 
infrastructure such as "Serverless"/AWS Lambdas.** 🚨

Snowflake IDs rely on the creation of a unique sequence number. On traditional deployment systems where each
"environment" necessitates exactly one physical or virtual server instance, atomic sequence number generation is 
only necessary between threads on each individual server. However, with modern deployment systems where a single
"environment" may consist of any number of physical or virtual servers, atomic locking must be ensured across
all server instances. This is only possible via a dedicated microservice, or via a shared cache system.

#### Usage
To use the ID generator of your choice on your Eloquent Model classes, simply add the corresponding trait
to your Model.

`Moves\Snowflake\Traits\EloquentTwitterSnowflakeId` for Twitter Snowflake

`Moves\Snowflake\Traits\EloquentSonyflakeId` for Twitter Snowflake

`Moves\Snowflake\Traits\EloquentModelflakeId` for Twitter Snowflake

or

`Moves\Snowflake\Traits\EloquentSnowflakeId` to specify your own custom ID Generator class

```
use Illuminate\Database\Eloquent\Model;
use Moves\Snowflake\Traits\EloquentTwitterSnowflakeId;

public class MyModel extends Model
{
    use EloquentTwitterSnowflakeId;
}
```

On `create`, your model class will automatically generate a unique ID and apply it to your model instance before
it is inserted into the database.

Note that JavaScript only supports up to 53-bit integers, and thus integer snowflake IDs will not be processed
as the correct value when retrieved as data from your API. As such, all of the included ID generation traits also
automatically add string casting on your model's key field so snowflake IDs are returned as strings when using 
`toArray` or `toJson`.

#### Overriding Default Sequence Number Generation
The supplied traits provide a default implementation for generating a unique sequence number, relying on Laravel's
Cache facade and atomic locking. However, you can override the default implementation by providing your own
function closure which returns a unique sequence number. 
See [Providing Your Own Generator](#Providing-Your-Own-Generator) for more details.

#### Providing Your Own Generator
To provide your own generator class, use the `Moves\Snowflake\Traits\EloquentSnowflakeId` trait, and implement the
following method header:
```
public function getSnowflakeGenerator(): ISnowflakeGenerator
{
    //TODO: Implement
}
```

You can build your own class which implements the `ISnowflakeGenerator` interface, or you can provide an instance
of one of the generators from this package with alternate values passed to the constructor.

For example, this is how you might provide your own sequence number generator function to the TwitterSnowflake
generator:
```
public function getSnowflakeGenerator(): ISnowflakeGenerator
{
    return new TwitterSnowflakeGenerator(
        $this->_getMachineId(),
        function (): int {
            // Closure containing your custom sequence number generation logic
        }
    );
}
```

### Direct Usage
This package is still usable outside of Eloquent model classes. Instantiate and call the appropriate generator
class wherever you need it:

```
$generator = new TwitterSnowflakeGenerator(
    $machineId,
    $sequenceNumberClosure
);

$id = $generator->generate;
```

You can also use the generators to parse an existing Snowflake ID into its component parts:
```
$parts = $generator->parse($snowflake);
// Expected Output:
// [timestamp] => 1634677016824 (ms)
// [machine] => 1
// [sequence] => 1
```


CHANGELOG.md
<!--- BEGIN HEADER -->
# Changelog

All notable changes to this project will be documented in this file.
<!--- END HEADER -->

## [3.0.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/2.0.3...3.0.0) (2023-04-10)

### ⚠ BREAKING CHANGES

* Upgrade to Laravel 10 ([9bf01e](https://github.com/vetmoves/com.moves.php.snowflake/commit/9bf01ec2f4d9c7759bc038b5078847e89c0b511c))


---

## [2.0.3](https://github.com/vetmoves/com.moves.php.snowflake/compare/2.0.2...2.0.3) (2023-02-27)

### Bug Fixes

* Use different keys for atomic lock and sequence value ([de5a45](https://github.com/vetmoves/com.moves.php.snowflake/commit/de5a458aab14f60f7d6f53a4b1cc905622913764))


---

## [2.0.2](https://github.com/vetmoves/com.moves.php.snowflake/compare/2.0.1...2.0.2) (2023-02-22)

### Bug Fixes

* Force release sequence cache lock ([e9b24a](https://github.com/vetmoves/com.moves.php.snowflake/commit/e9b24ac91cfc91ba76a4305f362bdb9bf05eb10a))


---

## [2.0.1](https://github.com/vetmoves/com.moves.php.snowflake/compare/2.0.0...2.0.1) (2023-02-22)

### Bug Fixes

* Parse snowflake sequence integer ([6b67af](https://github.com/vetmoves/com.moves.php.snowflake/commit/6b67afbda47a0cdf65f23c724cbe1b3ba1eee513))


---

## [2.0.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.8.0...2.0.0) (2022-09-07)

### ⚠ BREAKING CHANGES

* Upgrade to PHP 8.1 and Laravel 9 ([dc2823](https://github.com/vetmoves/com.moves.php.snowflake/commit/dc282320f416cf6e53d1eadacb15e4c1e645bf8f))


---

## [1.8.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.7.0...1.8.0) (2022-04-14)
### Features

* Allow null values for isSnowflake check, add isModel helper ([7b440a](https://github.com/vetmoves/com.moves.php.snowflake/commit/7b440aa29471cb2909a7d5c562e8d43d8b30dfa6))


---

## [1.7.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.6.2...1.7.0) (2022-04-14)
### Features

* Add isSnowflake helper function ([28525d](https://github.com/vetmoves/com.moves.php.snowflake/commit/28525d02b1ce4a67994e6b67e7a2cb799882b580))


---

## [1.6.2](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.6.1...1.6.2) (2022-03-16)
### Bug Fixes

* Fix bug where model key name is not included in guessSnowflakeFields ([f73767](https://github.com/vetmoves/com.moves.php.snowflake/commit/f73767fd43fc5768bfaba8716473b7066d2f383f))


---

## [1.6.1](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.6.0...1.6.1) (2022-03-16)
### Bug Fixes

* Use snake case for checking snowflake key suffix ([0679e1](https://github.com/vetmoves/com.moves.php.snowflake/commit/0679e1cffe72f12c51ef3de37838cbbd43200e0c))


---

## [1.6.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.5.0...1.6.0) (2022-03-16)
### Features

* Handle snowflake casting in Eloquent getCasts method, not on trait initialization ([2b7a94](https://github.com/vetmoves/com.moves.php.snowflake/commit/2b7a941ca2544e692f14dacade8b9a1ee76ed902))


---

## [1.5.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.4.0...1.5.0) (2022-03-16)
### Features

* Add automatic string casting on id fields ([296d43](https://github.com/vetmoves/com.moves.php.snowflake/commit/296d43e924d6d62c7a3f5fa379823c2d71fc0d2b))


---

## [1.4.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.3.0...1.4.0) (2022-02-21)
### Features

* Add class trait filter to Modelflake map command ([82cba2](https://github.com/vetmoves/com.moves.php.snowflake/commit/82cba2237685fb2993b0fb46b0146fbefbbc3fca))


---

## [1.3.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.2.0...1.3.0) (2021-12-21)
### Features

* Add override functions for generator traits ([edc9be](https://github.com/vetmoves/com.moves.php.snowflake/commit/edc9beda9ecd8ab94a3bd259a68183b5c0d17af8))


---

## [1.2.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.1.0...1.2.0) (2021-12-21)
### Features

* Add generateSnowflakeId function ([269fa6](https://github.com/vetmoves/com.moves.php.snowflake/commit/269fa6540a20b5d51f8599c215fffa8a4894eba0))


---

## [1.1.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.0.1...1.1.0) (2021-12-17)
### Features

* Add model function to parse generated ID ([1a232c](https://github.com/vetmoves/com.moves.php.snowflake/commit/1a232c99d7f760b6075394af8553c706e93f03ce))


---

## [1.0.1](https://github.com/vetmoves/com.moves.php.snowflake/compare/1.0.0...1.0.1) (2021-12-17)
### Bug Fixes

* Use full class name for Modelflake config ([50c9a1](https://github.com/vetmoves/com.moves.php.snowflake/commit/50c9a12e76f9f15bfa9dff46f5505c27c65e92ea))


---

## [1.0.0](https://github.com/vetmoves/com.moves.php.snowflake/compare/0.0.0...1.0.0) (2021-12-02)
### Features

* Implement Twitter Snowflake, Sonyflake, and Modelflake Generators
* Implement artisan command to automatically generate Modelflake mapping config

### Continuous Integration

* Configure conventional commit enforcement
* Configure automatic changelog

---



src/Generators/TwitterSnowflakeGenerator.php
<?php

namespace Moves\Snowflake\Generators;

use Closure;
use DateTime;
use DateTimeInterface;
use Moves\Snowflake\Contracts\ISnowflakeGenerator;
use Moves\Snowflake\Exceptions\SnowflakeBitLengthException;

/**
 * Class TwitterSnowflakeGenerator
 *
 * Implementation of Twitter Snowflake ID Generator
 * 63 bits total (64 - 1 sign bit)
 * 41 bits timestamp (unit ms), 10 bits machine id, 12 bits sequence number
 *
 * @see https://blog.twitter.com/engineering/en_us/a/2010/announcing-snowflake
 * @see https://developer.twitter.com/en/docs/twitter-ids
 */
class TwitterSnowflakeGenerator implements ISnowflakeGenerator
{
    //region Configuration
    /** @var int Allocated bits for timestamp */
    const BITS_TIMESTAMP = 41;

    /** @var int
     * Minimum number of bits that would be reasonable to represent a snowflake timestamp.
     * Used for determining if an integer is likely to be a snowflake ID.
     */
    const MIN_BITS_TIMESTAMP = 16;

    /** @var int Allocated bits for machine id */
    const BITS_MACHINE = 10;

    /** @var int Allocated bits for sequence number */
    const BITS_SEQUENCE = 12;

    /** @var string Default epoch start date */
    const DEFAULT_EPOCH = '1970-01-01 00:00:00 +0000 UTC';

    /** @var int Multiplier from microseconds to timestamp unit ms */
    const TIMESTAMP_MULTIPLIER = 1000;
    //endregion

    //region Constants
    public const SEQUENCE_MASK = (PHP_INT_MAX >> (self::BITS_TIMESTAMP + self::BITS_MACHINE));

    public const MACHINE_MASK =  (PHP_INT_MAX >> self::BITS_TIMESTAMP)
        ^ self::SEQUENCE_MASK;

    public const TIMESTAMP_MASK = PHP_INT_MAX
        ^ (self::MACHINE_MASK | self::SEQUENCE_MASK);
    //endregion

    //region Attributes
    /** @var int Machine ID */
    protected int $machineId;

    /** @var Closure Function to return unique sequence number */
    protected Closure $sequenceGenerator;

    /** @var int Epoch start timestamp */
    protected int $epochTimestamp;
    //endregion

    /**
     * TwitterSnowflakeGenerator constructor.
     * @param int $machineId Unique Machine ID
     * @param Closure $sequenceGenerator Function for generating unique sequence number
     * @param DateTimeInterface|null $epoch Optional epoch start time
     */
    public function __construct(int $machineId, Closure $sequenceGenerator, DateTimeInterface $epoch = null)
    {
        $this->machineId = $machineId;
        $this->sequenceGenerator = $sequenceGenerator;

        $epochDateTime = ($epoch ?? new DateTime(static::DEFAULT_EPOCH));
        $this->epochTimestamp = $epochDateTime->getTimestamp() * static::TIMESTAMP_MULTIPLIER;
    }

    /**
     * @inheritDoc
     */
    public function generate(): int
    {
        return ($this->getTimestampBits() << (static::BITS_MACHINE + static::BITS_SEQUENCE))
            | ($this->getMachineBits() << (static::BITS_SEQUENCE))
            | $this->getSequenceBits();
    }

    /**
     * @inheritDoc
     */
    public function isSnowflake(?int $snowflake): bool
    {
        return !is_null($snowflake) &&
            log($snowflake) > (static::MIN_BITS_TIMESTAMP + static::BITS_MACHINE + static::BITS_SEQUENCE);
    }

    /**
     * @inheritDoc
     */
    public function parse(int $snowflake): array
    {
        return [
            'timestamp' => $this->parseTimestampBits($snowflake),
            'machine' => $this->parseMachineBits($snowflake),
            'sequence' => $this->parseSequenceBits($snowflake)
        ];
    }

    public function getEpochTimestamp(): int
    {
        return $this->epochTimestamp;
    }

    //region Generate Helpers
    /**
     * @throws SnowflakeBitLengthException
     * @return int The timestamp
     */
    protected function getTimestampBits(): int
    {
        $timestamp = intval(microtime(true) * static::TIMESTAMP_MULTIPLIER) - $this->epochTimestamp;

        if ($timestamp >= (1 << static::BITS_TIMESTAMP) || $timestamp < 0) {
            throw new SnowflakeBitLengthException(
                'Timestamp',
                static::BITS_TIMESTAMP,
                $timestamp
            );
        }

        return $timestamp;
    }

    /**
     * @throws SnowflakeBitLengthException
     * @return int The machine ID
     */
    protected function getMachineBits(): int
    {
        if ($this->machineId >= (1 << static::BITS_MACHINE) || $this->machineId < 0) {
            throw new SnowflakeBitLengthException(
                'Machine ID',
                static::BITS_MACHINE,
                $this->machineId
            );
        }

        return $this->machineId;
    }

    /**
     * @throws SnowflakeBitLengthException
     * @return int The sequence number
     */
    protected function getSequenceBits(): int
    {
        $sequence = ($this->sequenceGenerator)();

        if ($sequence >= (1 << static::BITS_SEQUENCE) || $sequence < 0) {
            throw new SnowflakeBitLengthException(
                'Sequence',
                static::BITS_SEQUENCE,
                $sequence
            );
        }

        return $sequence;
    }
    //endregion

    //region Parse Helpers
    protected function parseTimestampBits(int $snowflake): int
    {
        return ($snowflake & static::TIMESTAMP_MASK) >> (static::BITS_MACHINE + static::BITS_SEQUENCE);
    }

    protected function parseMachineBits(int $snowflake): int
    {
        return ($snowflake & static::MACHINE_MASK) >> static::BITS_SEQUENCE;
    }

    protected function parseSequenceBits(int $snowflake): int
    {
        return $snowflake & static::SEQUENCE_MASK;
    }
    //endregion
}
